### PR TITLE
dq-ReviewBackend GET reviews needing moderation, PUT endpoints to update a user's review and moderate a review

### LINKED ISSUE #15: Dining Commons Reviews Backend - Controller method to get reviews awaiting approval plus tests
format: GET /api/reviews/needsmoderation

Gets all reviews with "Awaiting approval".

### PR DESCRIPTION
**Pull request by st4lemon**
Closes #15 #18  (combined into one pull request to make reviewing easier)

Go to: https://dining-st4lemon.dokku-13.cs.ucsb.edu

- Three new endpoints exist: POST /api/reviews/reviewer, POST /api/reviews/moderator, GET /api/reviews/needsmoderation:
diningmoderate
diningupdate
dininggetmoderation

- [ ] GET /api/reviews/needsmoderation displays all reviews needing moderation (if none exist, post a new review and verify it shows up).
- [ ] POST /api/reviews/moderator allows an admin user to approve/reject a review, and add comments. 
- [ ] Moderating a review correctly uses the moderator's user id as ModID.
- [ ] Moderated reviews do not show up in GET /api/reviews/needsmoderation.
- [ ] POST /api/reviews/reviewer allows a user to edit a review's star rating and review text.
- [ ] Editing an approved/rejected review resets the approval status to "Awaiting Approval", and deletes mod id and comments.
- [ ] All endpoints update lastEditedDate with the current time.

### CODE REVIEW

**Review by pconrad — APPROVED** [review]
LGTM!  Will test on dokku

File: src/main/java/edu/ucsb/cs156/dining/controllers/ReviewsController.java [hunk 1]
```diff
@@ -109,5 +122,62 @@ public Review postReview(
...
+    @PutMapping("/moderator")
+    public Review moderateReview(
+            @Parameter(name="id") @RequestParam Long id,
+            @RequestBody @Valid Review incoming) {
+
+
+        Review review = reviewsRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(Review.class, id));
+
+        review.setStatus(incoming.getStatus());
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/ReviewsController.java:173]: There should probably be some validation on this field—i.e. check whether the value is one of:
"Awaiting Moderation", "Approved", "Rejected", and if it isn't, throw an `IllegalArgumentException`.

**Review by pconrad — CHANGES_REQUESTED** [review]
This is really good work!  Just one small improvement to suggest...

File: src/main/java/edu/ucsb/cs156/dining/controllers/ReviewsController.java (hunk 1 above)
Comment (st4lemon) [reply src/main/java/edu/ucsb/cs156/dining/controllers/ReviewsController.java:188]: finished, with additional testing/methods for handling IllegalArgumentException.

File: src/main/java/edu/ucsb/cs156/dining/controllers/ReviewsController.java [hunk 2]
```diff
@@ -109,5 +136,66 @@ public Review postReview(
...
+            @Parameter(name="id") @RequestParam Long id,
+            @RequestBody @Valid Review incoming) {
+
+
+        Review review = reviewsRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(Review.class, id));
+        
+        if( !incoming.getStatus().matches("Approved|Awaiting Approval|Rejected") ) {
+            throw new IllegalArgumentException("Status must be 'Awaiting Moderation', 'Approved', or 'Rejected'");
+        }
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/ReviewsController.java:189]: Can we make the error message and the status consistent?  Even better if you can make that DRY.

Maybe something like: 
```java
ArrayList<String> legalStatusValues = new ArrayList<String>(
     Arrays.asList("Awaiting Moderation","Approved","Rejected");
```     

which you then use to check, as well as use in the error message.

Note that you could use:

```
"Status must be one of: " + legalStatusValues
```

and the toString() method would convert it nicely.

For the check, you could use:

```java
 if (!legalStatusValues.contains(incoming.getStatus()))

**Review by pconrad — CHANGES_REQUESTED** [review]
There's an inconsistency between the error checking and the error messages.

I suggest that you make it DRY so that this never happens... see below

File: src/main/java/edu/ucsb/cs156/dining/controllers/ReviewsController.java (hunk 2 above)
Comment (st4lemon) [reply src/main/java/edu/ucsb/cs156/dining/controllers/ReviewsController.java]: fixed

**Review by pconrad — APPROVED** [review]
LGTM!

**Pull request merged**